### PR TITLE
Renaming of Jobs by Pattern

### PR DESCRIPTION
**Pull request by wolfs**
Implementation of the possibility to configure a job so that on rename the existing Job is also renamed.

We give our job some index numbers to have a better ordering in the UI. These index number change when jobs are added or removed. In order not to loose the build history and not having to decide what to do with the old jobs this feature would be really handy.

Please comment if I tried implementing it the right way. If it is OK, then I would add some documentation.

### CODE REVIEW

File: job-dsl-plugin/build.gradle [hunk 1]
```diff
@@ -48,12 +52,12 @@ jenkinsPlugin {
...
 dependencies {
     compile project(':job-dsl-core')
-    optionalJenkinsPlugins 'org.jenkins-ci.plugins:cloudbees-folder:4.2@jar'
+    optionalJenkinsPlugins 'org.jenkins-ci.plugins:cloudbees-folder:4.4@jar'
     optionalJenkinsPlugins 'org.jenkins-ci.plugins:credentials:1.9.4@jar'
     optionalJenkinsPlugins 'org.jenkins-ci.plugins:vsphere-cloud:1.1.11@jar'
     optionalJenkinsPlugins 'org.jenkins-ci.plugins:config-file-provider:2.7@jar'
     jenkinsTest 'org.jenkins-ci.plugins:ant:1.2@jar' // see JENKINS-17129
-    jenkinsTest 'org.jenkins-ci.main:maven-plugin:1.509.3@jar'
+    jenkinsTest "org.jenkins-ci.main:maven-plugin:${jenkinsVersion}@jar"
```
Comment (daspilker) [inline job-dsl-plugin/build.gradle]: After 1.534 the maven-plugin got it's own release cycle, so there is no version 1.554.1. See http://repo.jenkins-ci.org/releases/org/jenkins-ci/main/maven-plugin/. Not sure why that didn't cause any problems?!?

I extracted the core update to #345. I will fix this pull request when merging.

**Comment by daspilker** [comment]
The documentation looks good!

I think we don't need the rename feature in StringJobManagement and FileJobManagement, there is IMHO no use case. StringJobManagement is just an internal test class and I would like to deprecate FileJobManagement. And both have no dedicated tests and I don't want to add any complex feature without tests.

I started a [discussion](https://groups.google.com/forum/#!msg/job-dsl-plugin/Z3pPQMyW-tY/MVljiFgfujYJ) on the mailing list about StringJobManagement and FileJobManagement. I will wait for comments until end of next week and then merge this pull request either with or without StringJobManagement and FileJobManagement.

**Comment by daspilker** [comment]
I deployed this pull request on Jenkins 1.565.3 and get a `java.lang.ClassNotFoundException: groovy.grape.GrabAnnotationTransformation` when running a seed job. The problem seems to be that the Groovy JAR is missing in the JPI file when building with Jenkins core 1.554.1. Some change in the Jenkins POMs must have triggered this. The same problem occurs for #345.

**Comment by daspilker** [comment]
I traced the problem down to a change between Jenkins 1.528 and 1.529: https://github.com/jenkinsci/jenkins/compare/jenkins-1.528...jenkins-1.529

IMHO the only relevant change is the jump from Groovy 1.8.5 to 1.8.9. In this pull request Groovy 1.8.6 is used for the Job DSL plugin and in #345 I updated to Groovy 1.8.9, but that didn't solve the problem.

**Comment by wolfs** [comment]
I rebased this change on #345.

**Comment by daspilker** [comment]
Ah sorry, I just force pushed #345 to fix the problem by introducing a force push configuration:
https://github.com/jenkinsci/job-dsl-plugin/pull/345/files#diff-50bbf6d5d2eba198671ca0f5ab0cc7c2R49

I think the problem didn't occur earlier because the Groovy version in the plugin (1.8.6) was newer than the version in core (1.8.5) which caused Gradle to include the artifact in the archive. With 1.554 the version in the plugin is older (1.8.6) or equals (1.8.9 in #345) the version in core, so Gradle does not include the artifact.

Using a `forceInclude` configuration does not seem to be very elegant to me. Do you know a better solution?

**Comment by wolfs** [comment]
Some investigation showed that the libs included in the jpi/hpi file are runtimeClasspath - providedRuntime. Since now the same groovy version is included in the provideRuntime by Jenkins it is not included any more in the jpi file.
On the long run (using this plugin from other plugins and probably having calls between them), we should not ship our own groovy jar. The ClassNotFoundException comes from the masked classes for the Grab support. See discussion at https://issues.jenkins-ci.org/browse/JENKINS-16361 and #51.
Like you already suggested in JENKINS-16361, including Ivy in Jenkins itself would solve the issue and relieve us from doing all the classloader magic.
Should I open a pull request for Jenkins-Core, including Ivy?

**Comment by wolfs** [comment]
See also discussion https://issues.jenkins-ci.org/browse/JENKINS-26192 and https://issues.jenkins-ci.org/browse/JENKINS-18349.

**Comment by daspilker** [comment]
Thanks for your analysis! Including Ivy in core will not solve the problem for now because we will have to wait for the PR to be merged and a LTS release containing the merge. And it will cause problems for plugins which use Ivy and want to use a specific (newer) version of Ivy, like the [Ivy Plugin](https://wiki.jenkins-ci.org/display/JENKINS/Ivy+Plugin), the [IvyTrigger Plugin](https://wiki.jenkins-ci.org/display/JENKINS/IvyTrigger+Plugin), the [Artifactory Plugin](https://wiki.jenkins-ci.org/display/JENKINS/Artifactory+Plugin) and probably others. So the chances that it may get merged are not that good. And I'm tempted to update Groovy to 2.x to get rid of some workarounds, see https://github.com/CoreMedia/job-dsl-plugin/compare/c5b2f1b...e620389. In that case the Job DSL plugin will have to include it's own version again. Unfortunately progress on the Groovy update in core (https://github.com/jenkinsci/jenkins/pull/1085) has stalled.

I would suggest to merge #345 as a short term solution. Open a pull request for including Ivy in core if you like, it would be an improvement for the Job DSL plugin. I will experiment with the update to Groovy 2.x and try to get update into core.

**Comment by wolfs** [comment]
Would groovy 2.x help for the Grab-Problem?

I still think, in order to get the extension point in place it would be very helpful to use the groovy brought by Jenkins. I even would favor getting rid of the Grab support altogether in favor of being able to add jars to the classpath via the plugin configuration.
If you want, we can have a discussion on this via google hangout or the likes.

**Comment by daspilker** [comment]
Groovy 2.x would help for IDE support, but maybe I should wait for an update in core.

You may be right, using a different version may cause problems for the extension point. I have not thought about that. I also thought about axing Grab support, but I think many people are using it. It can be replaced by the classpath option. The setup is a little bit more complex as it requires a build step if you don't want to put any JARs in SCM, but that can be documented. Can you start a discussion on the mailing list about Grab support? I will second the removal.

**Comment by wolfs** [comment]
For going forward with this pull request we can just stick to the fix in #345 and leave everthing as is. I also suppose it makes sense to merge #345 and keep the current behaviour.
If we deprecate @Grab support this probably should be another pull request.

**Comment by wolfs** [comment]
See discussion: https://groups.google.com/forum/#!topic/job-dsl-plugin/EG6eqQYYI7M

**Comment by wolfs** [comment]
Rebased to the updated #345

**Comment by daspilker** [comment]
Finally, after quiet some yak shaving, the update to 1.554 succeeded. I will rebase your changes to the current master when merging.

After the discussion on the mailing list, I will keep FileJobManagement. But I think it's not necessary to add support for renaming to FileJobManagement and StringJobManagement. I will simplify this when merging if you do not object.

**Comment by daspilker** [comment]
I rebased your changes to master and merged them manually. I did some refactoring and added some test cases with LookupStrategy.SEED_JOB. I had to fix the matching for relative names. I'm pretty sure I missed some edge cases.
